File: 400/483 - Smallest Good Base.java
class Solution {
    public String smallestGoodBase(String n) {
        long num = Long.valueOf(n), sum, left, right, mid;
        int i, j;
        
        for (i = (int)(Math.log(num + 1) / Math.log(2)); i >= 2; --i) {
            left = 2;
            right = (long)Math.pow(num, 1.0 / (i - 1)) + 1;
            
            while (left < right) {
                mid = left + (right - left) / 2;
                sum = 0;
                
                for (j = 0; j < i && sum < num; ++j)
                    sum = sum * mid + 1;
                
                if (sum == num)
                    return Long.toString(mid);
                else if (sum < num)
                    left = mid + 1;
                else
                    right = mid;
            }
        }
        
        return Long.toString(num - 1);
    }
}

File: csharp/483.smallest-good-base.cs
/*
 * @lc app=leetcode id=483 lang=csharp
 *
 * [483] Smallest Good Base
 */

using System;

// @lc code=start
public class Solution {
    public string SmallestGoodBase(string n) {
        long num = long.Parse(n);
        for (long length = (long) Math.Log(num + 1, 2); length >= 2; --length) {
            long left = 2, right = 1 + (long) Math.Pow(num, 1 / (double) (length - 1));
            while (left < right) {
                long sum = 0, mid = left + (right - left) / 2;
                for (long i = 0; i < length; ++i) {
                    sum = 1 + sum * mid;
                }
                if (sum == num) {
                    return mid.ToString();
                } else if (sum < num) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        return (num - 1).ToString();
    }
}
// @lc code=end

